src/shelly_hap_rgb.cpp

```diff
@@ -324,48 +361,91 @@ void RGBWLight::SaveState() {
 }
 
 Status RGBWLight::SetState(const std::string &state_json) {
-  bool state;
-  int brightness, hue, saturation;
+  bool state = false;
+  int brightness = 0, hue = 0, saturation = 0;
+
   json_scanf(state_json.c_str(), state_json.size(),
              "{state: %B, brightness: %d, hue: %d, saturation: %d}", &state,
              &brightness, &hue, &saturation);
 
-  if (cfg_->state != state) {
-    cfg_->state = state;
-    dirty_ = true;
-  }
-  if (cfg_->brightness != brightness) {
-    cfg_->brightness = brightness;
-    dirty_ = true;
+  if (hue < 0 || hue > 360) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT, "invalid hue: %d (only 0-360)",
+                        hue);
   }
-  if (cfg_->hue != hue) {
-    cfg_->hue = hue;
-    dirty_ = true;
-  }
-  if (cfg_->saturation != saturation) {
-    cfg_->saturation = saturation;
-    dirty_ = true;
+
+  if (saturation < 0 || saturation > 100) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT,
+                        "invalid saturation: %d (only 0-100)", saturation);
   }
 
-  if (dirty_) {
-    SetOutputState("RPC");
+  if (brightness < 0 || brightness > 100) {
+    return mgos::Errorf(STATUS_INVALID_ARGUMENT,
+                        "invalid brightness: %d (only 0-100)", brightness);
   }
 
+  UpdateOnOff(state, "RPC");
```

Comment:
if `state` is not provided, it goes to off (the default).
but what if i only want to change brightness?
if you look at other SetState implementations, they only change things that were requested. meaning, if there was no `state` key in the JSON, state should remain the same. it's a bit tricky with bools, you need to use an int8_t as the variable to use "not set" as default - see e.g. "in_inverted" in SetConfig method.
same applies to hue, saturation and brightness - if not provided, they should remain the same. you can use negative defaults for that.